<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Transcriptomics Analysis using R</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-M3MJYFFV49"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-M3MJYFFV49');
</script>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">About the course</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="preCourse.html">Pre-course</a>
</li>
<li>
  <a href="Configure.html">System configuration</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Exercises
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="chapter1.html">Chapter 1</a>
    </li>
    <li>
      <a href="chapter2.html">Chapter 2</a>
    </li>
  </ul>
</li>
<li>
  <a href="Assignment.html">Assignment</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Transcriptomics Analysis using R</h1>

</div>


<p>Welcome to the <strong>Next generation sequencing (NGS) and
bioinformatics 3BL008 BL00F HT2024</strong> course. This website aimed
to provide you with comprehensive resources for mastering
transcriptomics data analysis using R. Here, youâ€™ll find detailed
syllabi outlining the course structure, along with practical R code
examples and analysis techniques essential for your learning
journey.</p>
<p><br></p>
<div id="course-overview" class="section level1">
<h1>Course overview</h1>
<p>This course is structured around a central research problem:
identifying the molecular mechanisms caused by SARS-CoV-2 infection.
From the outset, students will be presented with this real-life
challenge and provided with actual RNA-seq data related to COVID-19. As
the course progresses, participants will learn and apply various
transcriptomics analysis techniques to solve this problem step by step.
Beginning with data normalization, students will prepare and standardize
the dataset, ensuring it is suitable for accurate analysis. They will
then employ dimensionality reduction methods, such as Principal
Component Analysis (PCA), to simplify the data and identify key patterns
and variations in gene expression.</p>
<p>Given that R is the primary tool for analysis in this course,
students will also receive training in the basics of R programming. This
foundational knowledge will enable them to use R-based packages
effectively throughout the course. Participants will start with an
introduction to R, covering essential commands, data manipulation, and
an overview of key R packages used in transcriptomics analysis. As they
progress, students will apply these R skills directly to their analysis,
enhancing their proficiency and confidence in using R for
bioinformatics.</p>
<p>At each stage of the course, students will build upon the previous
step, moving from differential expression analysis, where they will use
R based tools to identify genes significantly altered by SARS-CoV-2
infection to gene-set enrichment analysis, where they will explore and
interpret the broader biological pathways and processes affected by the
virus. Every practical session is directly tied to solving the research
problem, ensuring that students not only understand the theoretical
concepts but also apply them to uncover the molecular changes associated
with COVID-19.</p>
<p>By the end of the course, participants will have systematically
solved the research problem, identifying key molecular mechanisms
altered by SARS-CoV-2 infection. They will have developed detailed
visualizations, such as heatmaps, volcano plots and pathway diagrams,
and interpreted their findings to draw meaningful biological
conclusions. This step-by-step approach, combined with R training,
ensures that students leave the course with both the knowledge and
hands-on experience needed to independently tackle similar research
problems in their future work.</p>
</div>
<div id="course-objectives" class="section level1">
<h1>Course Objectives</h1>
<p>By the end of this course, participants will be able to:</p>
<ol style="list-style-type: decimal">
<li><strong>Understand the fundamentals of transcriptomics</strong> and
its significance in biological research.</li>
<li><strong>Use R for bioinformatics</strong> by gaining proficiency in
basic R programming.</li>
<li><strong>Perform downstream analysis</strong> of RNA-seq data,
including quality control, normalization, and differential expression
analysis.</li>
<li><strong>Interpret and visualize data</strong> using advanced
bioinformatics tools and methods.</li>
<li><strong>Apply functional enrichment analysis</strong> to uncover
biological pathways and mechanisms affected in disease states, with a
specific focus on COVID-19.</li>
<li><strong>Present and communicate findings</strong> effectively,
utilizing visualization techniques to support biological
interpretations.</li>
</ol>
</div>
<div id="course-structure" class="section level1">
<h1>Course Structure</h1>
<div id="module-1-introduction-to-transcriptomics"
class="section level2">
<h2>Module 1: Introduction to Transcriptomics</h2>
<p><strong>What is Transcriptomics?</strong>: Overview of
transcriptomics and its role in understanding gene expression.The
importance of transcriptomics in everyday life. Types of transcriptomics
data and technologies used to generate it. The Power of Transcriptomics
in mechanistic studies highlighting its role in identifying mechanisms
or processes that drive a specific biological phenomenon or disease, and
specifically to uncover the detailed molecular, cellular, and
physiological pathways involved, providing insights into how and why
certain effects or outcomes occur.</p>
</div>
<div id="module-2-the-research-problem" class="section level2">
<h2>Module 2: The Research Problem</h2>
<p><strong>Defining the Research Problem</strong>: Overview of the
specific research question to be addressed in the course.</p>
<p><strong>Study Design and Data Overview</strong>: Explanation of the
study design, including sample collection, experimental conditions, and
data acquisition. Introduction to the dataset: structure, content, and
format. Discussion on the challenges and limitations of the dataset.</p>
</div>
<div id="module-3-basics-of-r" class="section level2">
<h2>Module 3: Basics of R</h2>
<p><strong>Getting Started with R</strong>: Introduction to R and
RStudio environment. Basic R syntax and commands. Data types, variables,
and operators in R. Importing and managing data in R.</p>
<p><strong>R for Bioinformatics</strong>: Introduction to Bioconductor
and other R packages used for transcriptomics analysis. Hands-on
exercises to solidify R basics in the context of biological data.</p>
</div>
<div id="module-4-transcriptomics-data-analysis-pipeline"
class="section level2">
<h2>Module 4: Transcriptomics Data Analysis Pipeline</h2>
<p><strong>Preprocessing and Quality Control</strong>: Data
normalization techniques. Performing and interpreting Principal
Component Analysis (PCA) for sample distribution estimation.</p>
<p><strong>Differential Expression Analysis</strong>: Introduction to
differential gene expression analysis. Implementing DESeq2/edgeR in R to
identify differentially expressed genes. Interpretation of results:
identifying significant genes and understanding fold changes.</p>
<p><strong>Functional Enrichment Analysis</strong>: Overview of gene-set
enrichment analysis. Conducting functional enrichment analysis to
uncover biological pathways. Tools and packages in R for functional
analysis.</p>
</div>
<div id="module-5-data-visualization-and-interpretation"
class="section level2">
<h2>Module 5: Data Visualization and Interpretation</h2>
<p><strong>Visualizing Results</strong>: Creating publication-quality
plots and visualizations in R (e.g., volcano plots, heatmaps).
Customizing and refining visual outputs for clarity and impact.</p>
<p><strong>Interpretation and Biological Insight</strong>:
Systematically draw meaningful biological conclusions from the
results.</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
